### ISSUE TITLE
Multi-arch builds on docker.io

### ISSUE THREAD

**Issue by scottmalkie** [istio/istio#26652]
**Describe the feature request**

With support for linux/arm64 binaries as of 1.6.x, would it be possible to convert the amd64-only images on dockerhub to multi-arch images, so they can be used on mixed-architecture clusters?

Much of the "getting started" and other documentation fails on mixed-architecture or arm64-only clusters since the dockerhub images are amd64-only.

**Describe alternatives you've considered**
Rebuilding and separately hosting multi-arch builds in other repositories.

[ ] Docs
[X] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[X] Test and Release
[X] User Experience
[X] Developer Infrastructure

**Additional context**

It's even possible to use docker's new buildx functionality to easily create and push multi-arch images/manifests.

Thanks!

**Comment by howardjohn** [comment]
Related: https://github.com/istio/istio/issues/21094

**Comment by morlay** [comment]
@scottmalkie 

could you try https://github.com/querycap/istio

Need your feedbacks in your mixed-architecture cluster.

**Comment by ecliptik** [comment]
I got arm64 images building and setup for mutli-arch on Dockerhub (see: https://github.com/istio/istio/issues/21094#issuecomment-702429990), but used a EC2 `a2.xlarge` instance to build Istio on ARM natively. I imagine it's much more work to get Istio builds to play nicely with `docker buildx` and a cross-compile.

Having multi-arch officially baked into Istio images would be great though, since our current plan is to build/host our own Istio images with multi-arch support, but that's unsustainable over time.

**Comment by morlay** [comment]
@howardjohn is any plan to support this? 

I'm glad to merge my work https://github.com/querycap/istio to istio

Here is something may block the process 

### istio envoy binaries

wasms is not care the cpu arch, so could be compiled only once. here only need do for istio/envoy.

* [ ] istio-envoy-build-env for multi-arch

`envoyproxy/envoy-build-ubuntu` upgrade to llvm-10, which not support to build for istio/envoy. 

bazel < 3.5 in istio/envoy. have to need a unofficial bazel to support arm https://github.com/querycap/istio/blob/master/build/istio-envoy-build-env/Dockerfile.istio-envoy-build-env#L33

it could be used like: (amd64 & arm64 host supported both)
```shell
docker run \
   -e=TEST_TMPDIR=/tmp/bazel \
   -v=/tmp/bazel:/tmp/bazel \
   -v=${PWD}:/go/src/istio/proxy \
   -w=/go/src/istio/proxy \
   docker pull ghcr.io/querycap/istio/istio-envoy-build-env:0.1.0 make build_envoy
```

* [ ] build istio/envoy on arm64 host

<del>travis have timeout issue.</del>
enovy team using azure pipelines.
i setup a self-hosted arm64 host (now 12cpu) in github workflow https://github.com/querycap/istio/actions?query=workflow%3Aistio-envoy

it will be very very slow. amd64 host with qemu for arm64.

could istio team setup a runner on arm64 host too?

### istio binaries

easier for go binaries. 

Just build go on amd64 for target archs (without qemu)

* [ ] multi-arch istio base image

https://github.com/GoogleContainerTools/distroless starts to support multi-arch. 
After "*-debian10" published, multi-arch istio base images could be published 

The iptable hack could be
```Dockerfile 
FROM gcr.io/distroless/cc-debian10:latest AS distroless-amd64
COPY --from=default /sbin/xtables-multi /sbin/iptables* /sbin/ip6tables* /sbin/ip /sbin/
COPY --from=default /usr/lib/x86_64-linux-gnu/xtables/ /usr/lib/x86_64-linux-gnu/xtables
COPY --from=default /usr/lib/x86_64-linux-gnu/ /usr/lib/x86_64-linux-gnu
COPY --from=default /etc/iproute2 /etc/iproute2

FROM gcr.io/distroless/cc-debian10:latest AS distroless-arm64
COPY --from=default /sbin/xtables-multi /sbin/iptables* /sbin/ip6tables* /sbin/ip /sbin/
COPY --from=default /usr/lib/aarch64-linux-gnu/xtables/ /usr/lib/aarch64-linux-gnu/xtables
COPY --from=default /usr/lib/aarch64-linux-gnu/ /usr/lib/aarch64-linux-gnu
COPY --from=default /etc/iproute2 /etc/iproute2

FROM distroless-${TARGET_ARCH} as distroless
```

**Comment by sdake** [comment]
@morlay The Istio project is not particularly setup to handle other architectures. That said, we do enable the building of istioctl on other architectures as many people run on various platforms.

Would you mind filing an experimental enhancement request in the enhancement repo?

Here is where you file an issue that has broad scope: https://github.com/istio/enhancements/issues

Cheers,
-steve

**Comment by morlay** [comment]
@sdake 

hope i describle clear.
https://github.com/istio/enhancements/issues/14

**Comment by scottmalkie** [comment]
>  The Istio project is not particularly setup to handle other architectures. 

@sdake What would the Istio project need to handle other architectures, including building multi-arch images as part of the release process? Happy to assist however possible. Thanks!

**Comment by howardjohn** [comment]
We need our CI to be able to run ARM. Right now we run on GKE which doesn't support ARM

**Comment by scottmalkie** [comment]
Thanks so much for the ultra-quick response @howardjohn! 
Would it be possible to use docker buildx (or a remote build server hosted elsewhere) to support arm64?
How are the arm64 binaries themselves (since 1.6.3) built today?

**Comment by howardjohn** [comment]
Building arm64 go binaries is simple, go can cross compile easily. The issues are:
* istio/proxy is not trivial to cross compile. It may be possible, but its not as simple as go at least and no one has invested effort into doing it.
* Once we do build them, we also need to test them. I have experimented with emulation and the performance was no where near enough to run our tests, so we would need real hardware

**Comment by morlay** [comment]
And one more issue:

* build env and runtime env  for `istio/envoy` for x86_64 now is using ubuntu 16, which is not possible for arm64. if want to support arm64 and amd64 both, tool chains need to upgrade too.

@scottmalkie Before offical supports, https://github.com/querycap/istio could be used instead. 
I try to make sure arm64 version images same as amd64 ones.

and since 1.10, we could be easy to set `--hub=querycap/istio` to use istio tools like `istioctl`.

@howardjohn if any plan to build multi-arch, I think i could apply arm64 marchines for istio team from company, even from huawei cloud.

**Comment by scottmalkie** [comment]
Thank you @morlay, appreciate your work on querycap/istio.

@howardjohn We are more than happy to provide Istio with Graviton hosts to build/test any arm64-related aspects of the project, as well as any arm64-related guidance or assistance in terms of compilation, testing, or debugging. Eager to help get multi-arch images out the door however possible. Thanks so much!

**Comment by AWSjswinney** [comment]
@howardjohn I managed to run a native build of istio/proxy using the build-tools-proxy container from istio/tools that I reworked to build for Arm64. I haven't figured out how to modify the istio build scripts to use the locally built proxy binary, though. Is there any documentation on that?

If you/someone is working on this, I'd like to know what I can do to help.

@morlay I modified the Dockerfile for the build-tools* containers to use `ubuntu:focal` instead of `ubuntu:xenial` and it appears to work, but I haven't done extensive testing.

**Comment by morlay** [comment]
@AWSjswinney can't upgrade `ubuntu` only

I failed to upgrade build-tools directly https://github.com/istio/tools/pull/1346

* some upstream tools not provide arm64 version
* runtime required
  * `glibc` version must be same of higher as build tools 
  * `iptable` must be locked for `envoy`. see context https://github.com/istio/istio/pull/28686#issuecomment-732530131


And I have to setup a simpler build env
https://github.com/querycap/istio/blob/master/build/istio-envoy-build-env/Dockerfile.istio-envoy-build-env
you could see, deps for aarch64 & x86_64 have some diffrences like https://github.com/querycap/istio/blob/master/build/istio-envoy-build-env/Dockerfile.istio-envoy-build-env#L14-L25

You could use my build env to build your custom envoy follow this
https://github.com/querycap/istio/blob/master/build/istio/Dockerfile.istio-envoy#L10-L28

and for arm64 version of distroless images , i have to use the copy hacks (iptables version is too high in debian10,  but glibc version is too low in debian9)
https://github.com/querycap/istio/blob/master/build/istio/Dockerfile.proxyv2#L12-L18

**Comment by howardjohn** [comment]
We did update the images to ubuntu:focal and debian10 distroless by the way. Not sure if that helps at all

**Comment by morlay** [comment]
is `envoy` of `istio/proxy` already support `iptables@1.8` and `iptables@1.6` both ?

I see envoyproxy still build in `ubuntu:bionic` https://github.com/envoyproxy/envoy-build-tools/blob/main/build_container/Dockerfile-ubuntu#L13

if it already supported. `glibc` and `iptables` issues wil be fixed, i am sure about this.
it will be good news to build multi arch images.

and the upsteams tools without arm64 binary is not used for compiling, so could let it amd64 only
I think i could continue my PR for build-tools https://github.com/istio/tools/pull/1346

which release will use the new build env, `release-1.11` ? @howardjohn

**Comment by morlay** [comment]
@AWSjswinney

Here some changes for build-env https://github.com/istio/tools/pull/1346/files

**Comment by morlay** [comment]
Actions:

- [ ] Multi-arch build-tools
     - [x] Dockerfile https://github.com/istio/tools/pull/1346
     - [ ] Release
     
 - [ ]  Istio-proxy Envoy multi-arch builds
      - [ ] Need an aarch64 host (12u) in PROW (could use nodeSelector to lock aarch64 builds with the multi-arch build-tools-proxy) (@AWSjswinney could help)
      - [ ] Release
      
 - [ ] Istio
      - [ ] updates dockerx working dir https://github.com/istio/istio/pull/33763
      - [ ] Downloading amd64/arm64 envoy
      - [ ] Release

**Comment by morlay** [comment]
Now could build on aarch64 host on master, following this Makefile https://github.com/querycap/istio/blob/build-origin/Makefile

official multi-arch release is needs PROW runner for aarch64

**Comment by howardjohn** [comment]
I had previously thought we jsut needed hardware but I think some of the images in https://github.com/kubernetes/test-infra/issues/16588 are required as well to be ported:

- bootstrap
-  sidecar
-  initupload
-  clonerefs

**Comment by morlay** [comment]
Those images could have arm64 version, but with `-arm64` suffix, we could merge them as multi-arch images

* sidecar https://console.cloud.google.com/gcr/images/k8s-prow/GLOBAL/sidecar
* initupload https://console.cloud.google.com/gcr/images/k8s-prow/GLOBAL/initupload
* clonerefs https://console.cloud.google.com/gcr/images/k8s-prow/GLOBAL/clonerefs

This not have one arm64

* bootstrap https://console.cloud.google.com/gcr/images/k8s-testimages/GLOBAL/bootstrap
  * with the small change could build a multi-arch one https://github.com/kubernetes/test-infra/pull/23044

@howardjohn 

Should we waiting for offical release multi-arch images ? 
Or build them under namespace `gcr.io/istio-testing`?

**Comment by howardjohn** [comment]
it would be ideal to have them officially release I think they are updated daily so it would make things complex to have our own.

**Comment by morlay** [comment]
@howardjohn 

Howerver If someone want to setup PROW for arm64 before official release,  
could use images here: 
https://github.com/orgs/querycap/packages?repo_name=k8s-prow-images

**Comment by craigbox** [comment]
Here's a way to get access to arm64 machines: https://github.com/WorksOnArm/equinix-metal-arm64-cluster

Please LMK if you'd like it applied for on behalf of the project.

**Comment by morlay** [comment]
@zackzhangkai it is blocked by k8s team.
This PR couldn't be pushed https://github.com/kubernetes/test-infra/pull/23044

**Comment by infa-ddeore** [comment]
@morlay we used your `querycap/istio` repo to build istio `1.10.x` multi-arch image, envoy proxy was using 3x more cpu compared to AMD version or it.

I am not expert but may be building envoy needs some tweaking as mentioned in the link given by @scottmalkie 
https://github.com/aws/aws-graviton-getting-started/blob/main/c-c++.md ?

**Comment by infa-ddeore** [comment]
> @infa-ddeore
> 
> Do you compile it on native aarch64 host ? with qemu or other emulator, will be very slow. Even with 12u aarch64 host, envoy proxy in `querycap/istio` still cost 1h+ from zero.

@morlay I complied on graviton ec2 instance with 16 cpu machine, actually the cpu usage I am talking about is with the istio ingress proxy not with the build process it self. 

Istio ingress (envoy) running on graviton takes 3x more cpu compared to the istio ingress (envoy) running on AMD. We think that there may be some compilation flags change required for ARM binary.

**Comment by morlay** [comment]
Updates.

Now ci setup blocked by https://github.com/kubernetes/test-infra/pull/24783

Releted: https://github.com/istio/istio/issues/35176

**Comment by howardjohn** [comment]
One other note - https://github.com/envoyproxy/envoy/issues/19535.

There are still some rough edges but that allows cross compiling . might be a viable path forward.

To officially ship ARM support we will still need ARM nodes for tests but it's plausible zig could be the build mechanism

**Comment by mjsabby** [comment]
[Azure Kubernetes Service now supports ARM64 in preview](https://azure.microsoft.com/en-us/updates/public-preview-arm64-agent-node-support-in-aks/) in case testing is needed for this effort.

**Comment by howardjohn** [comment]
Aside from getting hardware to run this on, here is what jobs we need:
* Build build tools. This can be trivial today if we just build two separate images. However, we should have some setup that runs two builds (-amd64 and -arm64) then joins them into a single multi-arch image.
  Two options I have thought of, only needed because prow doesn't support complex flows:
  * In each job, append the arch to the manifest. This assumes we can concurrently modify the manifest (since we are racing across 2 jobs to create+ update the manifest)
  * In another job (this would likely be common-files, as it is the next step after istio/tools) we do the same, but a single job is creating the manifest instead of two
* Build envoy. This is straightforward, we just run two jobs (arm64 and amd64). We don't need to join.
  * Needs some simple modifications to the build script to publish to some architecture specific location though
* istio/istio.
  * Need minor change to pull correct envoy (see above), then I think running tests on arm node will Just Work
  * For building, I just added proper multi-arch builds in https://github.com/istio/istio/pull/39278 (not merged). We should cross compile for releases.
    * This probably needs slight tweak to get the correct envoy as well
* istio/release-builder: I think this is the same as above, we will cross compile. Just need to pass some argument to use both architectures. We can/should probably add a validation to the tests to ensure it builds correctly

**Comment by diranged** [comment]
@howardjohn Regarding the docker build process... most of our images internally are built with `docker buildx` ... but for the more complex and slower builds, we do the "build on native hardware" approach. It's pretty simple in CircleCI at least (and I am sure this translates well to Github Actions too):

Here's an example `.circleci/config.yaml`:

```yaml
version: 2.1

jobs:
  build-amd64: &build-default
    machine:
      image: ubuntu-2004:202101-01
      docker_layer_caching: true
    steps:
      - checkout
      - run: make ecr_login
      - run: make docker_build_for_platform_and_push

  build-arm64:
    <<: *build-default
    resource_class: arm.large

  build:
    docker:
      - image: cimg/python:3.8.7
    steps:
      - checkout
      - run: make venv
      - run: make ecr_login
      - run: make docker_create_manifest_and_push
      - run: |
          if [ "${CIRCLE_BRANCH}" == "main" ]; then
            make docker_create_latest_manifest_and_push
          fi
workflows:
  version: 2
  workflow:
    jobs:
      - build:
          context: staging
          requires:
            - build-amd64
            - build-arm64
          filters:
            tags:
              only: /^((([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)$/

      - build-arm64:
          context: staging
          filters:
            tags:
              only: /^((([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)$/

      - build-amd64:
          context: staging
          filters:
            tags:
              only: /^((([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)$/
```

Our `Makefile` looks something like this (this is stripped a bit...)
```
DOCKER_ARCH := $(shell docker info -f '{{json .Architecture }}' 2>/dev/null)
CPU_ARCHS ?= x86_64,aarch64

.PHONY: docker_build
docker_build:
	docker build Dockerfile \
		-t $(DOCKER_IMAGE) \
		-t $(DOCKER_IMAGE):$(DOCKER_TAG) \
		-t $(DOCKER_NAME) \
		-t $(DOCKER_FQDN)

.PHONY: docker_push
docker_push: docker_tag
	docker push $(DOCKER_FQDN)

.PHONY: docker_build_for_platform_and_push
docker_build_for_platform_and_push: check_platform
	DOCKER_FQDN=$(DOCKER_FQDN)-$(DOCKER_ARCH) $(MAKE) docker_build docker_push

.PHONY: docker_create_manifest_and_push
docker_create_manifest_and_push:
	DOCKER_CLI_EXPERIMENTAL=enabled && \
	docker manifest create $(DOCKER_MANIFEST_FQDN) \
		$(shell IFS="," && for a in $$(echo $(CPU_ARCHS)); do amend="$$amend --amend $(DOCKER_FQDN)-$$a"; done && echo $$amend) && \
	docker manifest push $(DOCKER_MANIFEST_FQDN)
```

The basic answer is what you already had .... two separate and parallel build steps that run on native hardware to build a "$image:$tag-$arch" image and they push it to your target repo of choice. Then a final followup step that builds the manifest and pushes the "$image:$tag".

**Comment by howardjohn** [comment]
Unfortunately prow is quite limited and cannot do dependencies between jobs which is why I was looking through the various hacky solutions

**Comment by diranged** [comment]
Ah. What is “prow” - I just figured you guys used either CIrcleCI, Github or AWS CodeBuild to do the builds (or some Google thingy).

**Comment by craigbox** [comment]
Good news, everyone!
https://cloud.google.com/blog/products/containers-kubernetes/gke-supports-new-arm-based-tau-t2a-vms

😉

**Comment by vishnu-anil** [comment]
We also need this. Please do prioritize

**Comment by akash123-eng** [comment]
This is much needed for our project also
Please prioritize

**Comment by howardjohn** [comment]
arm64 support has official landed! Thank you to everyone that helped get this over the finish line, including @morlay , @TrevorTaoARM , @ericvn , @stewartbutler and many more.

Currently support is on the `master` branch. To try it out, follow: https://github.com/istio/istio/wiki/Dev-Builds#dev-release-information. Usage is the same - just `istioctl install` and you should be good to go.

Support will be included in the 1.15 release coming out in ~1 month.

If you have issues, please open a new issue rather than commenting here.

**Comment by TrevorTaoARM** [comment]
> arm64 support has official landed! Thank you to everyone that helped get this over the finish line, including @morlay , @TrevorTaoARM , @ericvn , @stewartbutler and many more.
> 
> Currently support is on the `master` branch. To try it out, follow: https://github.com/istio/istio/wiki/Dev-Builds#dev-release-information. Usage is the same - just `istioctl install` and you should be good to go.
> 
> Support will be included in the 1.15 release coming out in ~1 month.
> 
> If you have issues, please open a new issue rather than commenting here.

@howardjohn It's really a great milestone. I would like to ask if the Istio arm CI is available now, or we still need to build a Istio Arm CI environment which can support various jobs as those of amd64.

**Comment by howardjohn** [comment]
No, we have a CI environment now which this is built on. Release+test are
automated alongside amd64 jobs

**Comment by TrevorTaoARM** [comment]
> No, we have a CI environment now which this is built on. Release+test are automated alongside amd64 jobs
> […](#)

Thanks for the update!
